### DISCUSSION TITLE
关于数据库版本对于用户和模型的使用量监控问题

### DISCUSSION THREAD

**Discussion by SoulGolf** [lobehub/lobehub#4936] [General Question] [answered]
我目前的场景是部署数据库版本后提供给了不少朋友间使用,同时用了不少的服务提供商,就会存在2个问题,不知道朋友们都喜欢使用哪个模型?或者哪个用户的使用量比较多?好了解哪个模型好用之类的,也方便重点进行维护

**Comment by SoulGolf** [comment]
目前支持的监测中,哪一个可以做到以上的要求?

**Reply by breakstring** [reply]
> 目前支持的监测中,哪一个可以做到以上的要求?

对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。

后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了

**Reply by SoulGolf** [reply]
> > 目前支持的监测中,哪一个可以做到以上的要求?
> 
> 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> 
> 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了

One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议,
朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆

**Reply by SpeedupMaster** [reply]
> > > 目前支持的监测中,哪一个可以做到以上的要求?
> > 
> > 
> > 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> > 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了
> 
> One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议, 朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆

感觉openrouter的模型挺全的，我现在一些开源模型都是用openrouter上的

**Reply by breakstring** [reply]
> > > 目前支持的监测中,哪一个可以做到以上的要求?
> > 
> > 
> > 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> > 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了
> 
> One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议, 朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆

需要用我的 Lobechat服务的朋友，大多对于模型的能力没啥喜好或者明确的认知。。。一般来说就是给他们啥他们用啥，估计好多人都不知道可以换模型。。。噗。。。。。
所以，我是在我的配置里写好我推荐他们用的，以及一些提示信息
例如。。。。。。

![image](https://github.com/user-attachments/assets/b0587122-965d-4bd4-beed-a057bd2c9962)

*之所以我推荐给他们用 Azuer 的是因为这个是我薅羊毛来的目前免费😂

**Reply by SoulGolf** [reply]
> > > > 目前支持的监测中,哪一个可以做到以上的要求?
> > > 
> > > 
> > > 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> > > 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了
> > 
> > 
> > One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议, 朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆
> 
> 需要用我的 Lobechat服务的朋友，大多对于模型的能力没啥喜好或者明确的认知。。。一般来说就是给他们啥他们用啥，估计好多人都不知道可以换模型。。。噗。。。。。 所以，我是在我的配置里写好我推荐他们用的，以及一些提示信息 例如。。。。。。
> 
> ![image](https://private-user-images.githubusercontent.com/14105911/393695155-b0587122-965d-4bd4-beed-a057bd2c9962.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3MjQ3NzgsIm5iZiI6MTczMzcyNDQ3OCwicGF0aCI6Ii8xNDEwNTkxMS8zOTM2OTUxNTUtYjA1ODcxMjItOTY1ZC00YmQ0LWJlZWQtYTA1N2JkMmM5OTYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDA2MDc1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1YjQ1YmZkYWZkNDcwNGNkNDQ4NDVmOWFmZjMzMTNjYzE1ZTZlMmQ5OWRkM2I0YTk0ZGI2YWE3OWUzNGYxYjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6HdjYc_8X8-DGzygzAZZvN0HjOGFevIWOi5QqjlPy7I)
> 
> *之所以我推荐给他们用 Azuer 的是因为这个是我薅羊毛来的目前免费😂

哈哈哈,我跟你是差不多的,我只把性价比高和能力强的模型放出来,白嫖了一堆免费模型

**Reply by SoulGolf** [reply]
> > > > 目前支持的监测中,哪一个可以做到以上的要求?
> > > 
> > > 
> > > 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> > > 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了
> > 
> > 
> > One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议, 朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆
> 
> 感觉openrouter的模型挺全的，我现在一些开源模型都是用openrouter上的

国外的开源模型基本都有免费额度用不完,例如:Together 和Groq的免费计划完全够用

**Reply by SpeedupMaster** [reply]
> > > > > 目前支持的监测中,哪一个可以做到以上的要求?
> > > > 
> > > > 
> > > > 对于前者的话，我觉得如果通过某一个 API网关/聚合型的工具来做一层代理，应该都能做到统计。例如 One-API, Cloudflare 的 AI Gateway, portkey 等等。
> > > > 后者。。。比较麻烦，要不你从那些聚合性网关里给你朋友分别生成 key 让他们自己在客户端配置。。。其实倒是没啥必要，如果只是给几个朋友用估计这个用量。。。没必要纠结了
> > > 
> > > 
> > > One-API我是在用的,接入了一部分,对于那种开源模型的提供商来说模型太多了也是不好维护的,而且也不是所有提供商都支持了openai协议, 朋友少确实也不怎么纠结,但是随着时间的流逝,有点好奇他们都喜欢用啥模型多😆
> > 
> > 
> > 感觉openrouter的模型挺全的，我现在一些开源模型都是用openrouter上的
> 
> 国外的开源模型基本都有免费额度用不完,例如:Together 和Groq的免费计划完全够用

开源模型还是用得少，主要还是用openai、google、Claude这三家的

**Comment by arvinxx** [comment]
基于 messages 表可以做一个统计。你可以让 claude sonnet3.5 来给你写 db sql 做查询的，比如我们现在 cloud 上的数据用量：

<img width="1263" alt="image" src="https://github.com/user-attachments/assets/7c138536-7aac-49d9-b67f-d2dabf7f4175">

表结构在这里：https://github.com/lobehub/lobe-chat/tree/main/src/database/schemas。 SQL 为：

```sql
SELECT 
    model,
    COUNT(*) as usage_count
FROM messages
WHERE 
    role = 'assistant' 
    AND model IS NOT NULL 
GROUP BY model
ORDER BY usage_count DESC;
```

**Reply by arvinxx** [reply, marked as answer]
我和我的助手对话：

![LobeChat_数据库解决方案专家_2024-12-09](https://github.com/user-attachments/assets/f51eeeca-e7fa-43bf-8ec2-7f22af144360)

**Reply by SoulGolf** [reply]
准备用metabase接入pg库实时可视化分析看看效果

**Reply by arvinxx** [reply]
不错，也是个思路

**Reply by m0eak** [reply]
> 准备用metabase接入pg库实时可视化分析看看效果

能出个教程吗

**Reply by SoulGolf** [reply]
> > 准备用metabase接入pg库实时可视化分析看看效果
> 
> 能出个教程吗

部署MetaBase配置上LobeChat的数据库地址就行了,根据上面给的SQL示例去改你要的可视化结果

**Reply by m0eak** [reply]
> > > 准备用metabase接入pg库实时可视化分析看看效果
> 
> > 
> 
> > 能出个教程吗
> 
> 
> 
> 部署MetaBase配置上LobeChat的数据库地址就行了,根据上面给的SQL示例去改你要的可视化结果

搞了一个，挺好用的，但是每个用户使用了哪些模型，好像没有找到对应的模块。

**Reply by arvinxx** [reply]
@m0eak 可以看 message 的统计。目前每条助手的 message 里有存用了哪个 provider 和哪个模型

**Reply by Matinal111** [reply]
![image](https://github.com/user-attachments/assets/79bc5d59-2dca-4542-a5ad-f02a94487c19)
按照上面的 SQL 示例用 Metabase 来做可视化分析确实不错，不过要基于 messages 表统计哪些用户使用了哪些模型好像不行，大佬能给个示例吗😂

**Reply by SoulGolf** [reply]
> ![image](https://private-user-images.githubusercontent.com/45492444/397586199-79bc5d59-2dca-4542-a5ad-f02a94487c19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ2NjE3OTUsIm5iZiI6MTczNDY2MTQ5NSwicGF0aCI6Ii80NTQ5MjQ0NC8zOTc1ODYxOTktNzliYzVkNTktMmRjYS00NTQyLWE1YWQtZjAyYTk0NDg3YzE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjIwVDAyMjQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MmJmYTYwM2I5MjlhOTgyZDU4MDZlY2ExZmZiNjk2ZWQyZGUxZDhkMzBlNmNjMTBjNjg5NTEwZDdiYjk2N2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZwIEh457TMn9431WP90xUBfcJgBS49Mj3PK7AXNLLx0) 按照上面的 SQL 示例用 Metabase 来做可视化分析确实不错，不过要基于 messages 表统计哪些用户使用了哪些模型好像不行，大佬能给个示例吗😂

SELECT
  user_id,model,
  COUNT(*) AS daily_usage
FROM
  messages
WHERE
  model IS NOT NULL
GROUP BY
  user_id,model;

**Reply by Matinal111** [reply]
近期 LobeChat 已经实现了使用量可视化分析，MetaBase 还是太占用服务器资源，唯一的问题就是不能看到其他用户的使用数据，不知道后续会不会改进

**Reply by SoulGolf** [reply]
> 近期 LobeChat 已经实现了使用量可视化分析，MetaBase 还是太占用服务器资源，唯一的问题就是不能看到其他用户的使用数据，不知道后续会不会改进

主要还是对其他用户的使用监测,后期如果能增加一个管理员专属界面就好了

**Reply by chung1912** [reply]
> > 近期 LobeChat 已经实现了使用量可视化分析，MetaBase 还是太占用服务器资源，唯一的问题就是不能看到其他用户的使用数据，不知道后续会不会改进
> 
> 主要还是对其他用户的使用监测,后期如果能增加一个管理员专属界面就好了

比如环境变量增加一个管理员角色，后台设置用户A为管理员，然后，在管理员的界面上，可以查看任何用户的数据统计，也可以查看所有用户或者部分多个用户的数据统计

**Reply by arvinxx** [reply]
未来的团队版/商业版/企业版会提供这方面的能力。管理类的功能就会是收费的了。
